src/module_3/evaluation_functions.py

```diff
@@ -0,0 +1,179 @@
+        best_threshold_index = valid_thresholds[0][0]
+    
+    best_threshold = thresholds[best_threshold_index]
+    print('Best threshold={:.3f}, {} on validation={:.2f}'.format(best_threshold, metric, performance_thresholds[best_threshold_index]))
+    return best_threshold
+
+def evaluate_best_threshold_test(model, threshold, X_test, y_test, metric='fscore'):
+    if hasattr(model, 'predict_proba'):
+        model_probs_test = model.predict_proba(X_test)
+        model_probs_test = model_probs_test[:, 1]
+    else:
+        model_probs_test = model.decision_function(X_test)
+        
+    if (metric == 'fscore'):
+        print('Fscore of the best model on test after optimizing threshold {:.2f}'.format(metrics.f1_score(y_test, threshold_classification(model_probs_test, threshold)) * 100.0))
+    elif (metric == 'recall_mean'):
+        print('Mean recall by classes of the best model on test after optimizing threshold {:.2f}'.format(metrics.roc_auc_score(y_test, threshold_classification(model_probs_test, threshold)) * 100.0))
+    elif (metric == 'fpr'):
+        print('FPR of the best model on test after optimizing threshold {:.2f}'.format(100 - metrics.recall_score(y_test, threshold_classification(model_probs_test, threshold), pos_label=1) * 100.0))
+    predictions_test_best_threshold = threshold_classification(model_probs_test, threshold)
+    print(metrics.confusion_matrix(y_test, predictions_test_best_threshold, labels=[1, 0]))
+
+def plot_validation_curve(train_scores, test_scores, param_range, ax=None, param='Param', score='Score', scale=None):
+    train_scores_mean = np.mean(-train_scores, axis=1)
+    train_scores_std = np.std(-train_scores, axis=1)
+    test_scores_mean = np.mean(-test_scores, axis=1)
+    test_scores_std = np.std(-test_scores, axis=1)
+    
+    if ax is None:
+        fig, ax = plt.subplots(1, 1, figsize=(8, 6))
+
+    ax.set_title("Validation Curve")
+    ax.set_xlabel(param)
+    ax.set_ylabel(score)
+        
+    lw = 2
+    if scale is None:
+        plot_fun = ax.plot
+    else:
+        plot_fun = ax.semilogx
+    plot_fun(param_range, train_scores_mean, label="Training score",
+                 color="darkorange", lw=lw)
+    ax.fill_between(param_range, train_scores_mean - train_scores_std,
+                     train_scores_mean + train_scores_std, alpha=0.2,
+                     color="darkorange", lw=lw)
+    plot_fun(param_range, test_scores_mean, label="Cross-validation score",
+                 color="navy", lw=lw)
+    ax.fill_between(param_range, test_scores_mean - test_scores_std,
+                     test_scores_mean + test_scores_std, alpha=0.2,
+                     color="navy", lw=lw)
+    ax.legend(loc="best")
+    
+    return ax
+
+def evaluate_classifier(clf, X, y):
+    prediction = clf.predict(X)
+    recall_mean = metrics.roc_auc_score(y, prediction) * 100
+    cm = metrics.confusion_matrix(y, prediction, labels=[1, 0])
+    recall = cm[0, 0] / cm[0, :].sum() * 100
```

Comment:
This is kinda wrong. It's the recall assuming threshold is 0.5. It doesn't need to be and it shouldn't be.